### ISSUE TITLE
Custom GCP image_id errors out `Head node fetch timed out. Failed to create head node.`

### ISSUE THREAD

**Issue by tokestermw** [skypilot-org/skypilot#2666]
Hi, I am having trouble setting a custom image on GCP. I built a new CUDA 11.8 image using the Google Cloud console, but whenever I set `image_id`, I get an error `Head node fetch timed out. Failed to create head node.`. Without setting the `image_id`, fetching the head node succeeds.

Skypilot config looks something like this.

```yaml
resources:
  cloud: gcp
  ...
  image_id: projects/PROJECT_NAME/global/images/deeplearning-vm-pytorch-2-cuda-118-us
```

And the error message, using `skypilot==1.0.0dev0`.

```shell
─────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────
I 10-05 17:16:17 optimizer.py:732] Considered resources (1 node):                                                           [85/1373]
I 10-05 17:16:17 optimizer.py:781] --------------------------------------------------------------------------------------------
I 10-05 17:16:17 optimizer.py:781]  CLOUD   INSTANCE        vCPUs   Mem(GB)   ACCELERATORS   REGION/ZONE   COST ($)   CHOSEN
I 10-05 17:16:17 optimizer.py:781] --------------------------------------------------------------------------------------------
I 10-05 17:16:17 optimizer.py:781]  GCP     a2-highgpu-8g   96      680       A100:8         us-west4      31.47         ✔
I 10-05 17:16:17 optimizer.py:781] --------------------------------------------------------------------------------------------
I 10-05 17:16:17 optimizer.py:781]
I 10-05 17:16:17 cloud_vm_ray_backend.py:1254] To view detailed progress: tail -n100 -f /root/sky_logs/sky-2023-10-05-16-59-42-328960
/provision.log
I 10-05 17:16:44 cloud_vm_ray_backend.py:1578] Launching on GCP us-west4 (us-west4-b)
I 10-05 17:18:56 cloud_vm_ray_backend.py:713] ====== stdout ======
2023-10-05 17:16:44,799 INFO commands.py:276 -- Cluster: axolotl-test
2023-10-05 17:16:45,316 INFO commands.py:353 -- Checking External environment settings
2023-10-05 17:16:56,708 INFO commands.py:654 -- No head node found. Launching a new cluster. Confirm [y/N]: y [automatic, due to --ye
s]
2023-10-05 17:16:56,708 INFO usage_lib.py:372 -- Usage stats collection is disabled.
No head node exists, need to create it.
2023-10-05 17:16:56,708 INFO commands.py:711 -- Acquiring an up-to-date head node
I 10-05 17:17:01 node.py:343] wait_for_compute_zone_operation: Waiting for operation operation-1696526217430-606fb4a8ac70a-2df84b90-a
fcc4c3e to finish...
2023-10-05 17:18:03,858 INFO commands.py:727 -- Launched a new head node
2023-10-05 17:18:03,858 INFO commands.py:731 -- Fetching the new head node

I 10-05 17:18:56 cloud_vm_ray_backend.py:716] ====== stderr ======
Dropping the empty legacy field head_node. head_nodeis not supported for ray>=2.0.0. It is recommended to removehead_node from the cl
uster config.
Dropping the empty legacy field worker_nodes. worker_nodesis not supported for ray>=2.0.0. It is recommended to removeworker_nodes fr
om the cluster config.
2023-10-05 17:18:56,606 PANIC commands.py:734 -- Head node fetch timed out. Failed to create head node.
Traceback (most recent call last):
  File "/tmp/skypilot_ray_up__ep8omu0.py", line 76, in <module>
    sdk.create_or_update_cluster('/root/.sky/generated/axolotl-test.yml', **{'no_restart': True})
  File "/root/venv3.8/lib/python3.8/site-packages/ray/autoscaler/sdk/sdk.py", line 38, in create_or_update_cluster
    return commands.create_or_update_cluster(
  File "/root/venv3.8/lib/python3.8/site-packages/ray/autoscaler/_private/commands.py", line 282, in create_or_update_cluster
    get_or_create_head_node(
  File "/root/venv3.8/lib/python3.8/site-packages/ray/autoscaler/_private/commands.py", line 734, in get_or_create_head_node
    cli_logger.abort(
  File "/root/venv3.8/lib/python3.8/site-packages/ray/autoscaler/_private/cli_logger.py", line 614, in abort
    raise exc_cls(msg)
click.exceptions.ClickException: Head node fetch timed out. Failed to create head node.
```

**Comment by concretevitamin** [comment]
Hello @tokestermw! It's possible this log has more info:
```
I 10-05 17:16:17 cloud_vm_ray_backend.py:1254] To view detailed progress: tail -n100 -f /root/sky_logs/sky-2023-10-05-16-59-42-328960
/provision.log
```

Previously, we've seen errors similar to this being caused by the image not having `rsync`/other tools installed. Are there any hints about this in `provision.log`?

**Comment by tokestermw** [comment]
thanks @concretevitamin I don't see anything super relevant. I also just checked the image, and it does have `rsync` installed.

```
2023-10-05 17:16:56,708 INFO commands.py:654 -- No head node found. Launching a new cluster. ESC[4mConfirm [y/N]:ESC[24m y ESC[2m[automatic, due to --yes]ESC[22m
2023-10-05 17:16:56,708 INFO usage_lib.py:372 -- Usage stats collection is disabled.
No head node exists, need to create it.
2023-10-05 17:16:56,708 INFO commands.py:711 -- ESC[36mAcquiring an up-to-date head nodeESC[39m
I 10-05 17:17:01 node.py:343] wait_for_compute_zone_operation: Waiting for operation operation-1696526217430-606fb4a8ac70a-2df84b90-afcc4c3e to finish...
2023-10-05 17:18:03,858 INFO commands.py:727 -- Launched a new head node
Traceback (most recent call last):
  File "/tmp/skypilot_ray_up__ep8omu0.py", line 76, in <module>
2023-10-05 17:18:03,858 INFO commands.py:731 -- ESC[36mFetching the new head nodeESC[39m
    sdk.create_or_update_cluster('/root/.sky/generated/axolotl-test.yml', **{'no_restart': True})
  File "/root/venv3.8/lib/python3.8/site-packages/ray/autoscaler/sdk/sdk.py", line 38, in create_or_update_cluster
    return commands.create_or_update_cluster(
  File "/root/venv3.8/lib/python3.8/site-packages/ray/autoscaler/_private/commands.py", line 282, in create_or_update_cluster
    get_or_create_head_node(
  File "/root/venv3.8/lib/python3.8/site-packages/ray/autoscaler/_private/commands.py", line 734, in get_or_create_head_node
    cli_logger.abort(
  File "/root/venv3.8/lib/python3.8/site-packages/ray/autoscaler/_private/cli_logger.py", line 614, in abort
    raise exc_cls(msg)
click.exceptions.ClickException: Head node fetch timed out. Failed to create head node.
```

**Comment by Michaelvll** [comment]
Thanks for reporting this @tokestermw! I am looking into the reason for this. Just curious, if you `sky launch -c cluster-name` again, will it normally setup the cluster?

**Comment by Michaelvll** [comment]
I can reproduce the issue with `sky launch --cloud gcp --gpus A100:8 --image-id projects/deeplearning-platform-release/global/images/pytorch-2-0-gpu-v20230822-ubuntu-2004-py310 --down --retry-until-up`. It seems the problem might be some flaky error from GCP. We can add some retry for this kind of failure.

**Comment by Michaelvll** [comment]
We now fixed the issue in #2669 ; ) It will be merged soon.

**Comment by Michaelvll** [comment]
We now fixed the issue in #2669 ; ) It will be merged soon. @tokestermw

**Comment by tokestermw** [comment]
Thank you! I did try again and I was able to train so it does look intermittent.

I wonder if it has anything to do with the image is created for us region but the gpu is in asia (or vice-versa)..

**Comment by Michaelvll** [comment]
Nice! Thanks for letting us know. Great to hear it works with retry. In #2669, we add the automatic retry, so no need to launch again.

> I wonder if it has anything to do with the image being created in the us region but the gpu is in asia (or vice-versa)...

We are not sure if the image on GCP has the concept of location, but we did observe a larger image could take longer to provision, which may trigger some timeout.
However, it seems more likely that GCP has some flaky issues with A100:8, e.g., we experienced a few times that the on-demand VM was deleted by the cloud after we got the IP of the VM. 

Just a FYI, if you are trying to get A100 resources `--retry-until-up` argument would be helpful to let SkyPilot keep trying to grab an available VM for you. ; )
